pandas/core/internals/blocks.py

```diff
@@ -801,10 +801,23 @@ def _replace_list(
 
         rb = [self if inplace else self.copy()]
         for i, (src, dest) in enumerate(pairs):
+            convert = i == src_len  # only convert once at the end
             new_rb: List[Block] = []
+            mask_pos = 0
             for blk in rb:
-                m = masks[i]
-                convert = i == src_len  # only convert once at the end
+                if blk.ndim == 1:
+                    m = masks[i]
+                else:
+                    # GH-39338: _replace_coerce can split a block, so we
+                    # need to keep track of where to index into the mask
+                    assert not isinstance(masks[i], bool)
+                    # error: Value of type "Union[ExtensionArray, ndarray, bool]"
+                    # is not indexable
+                    m = masks[i][
+                        mask_pos : mask_pos + blk.shape[0]
+                    ]  # type: ignore[index]
```

Comment:
i think if you define `mib = masks[i]` or something that might make this type:ignore unnecessary

Reply:
good call!